ee/hogai/graph/filter_options/nodes.py

```diff
@@ -0,0 +1,321 @@
+        toolkit = FilterOptionsToolkit(self._team)
+        intermediate_steps = state.intermediate_steps or []
+        action, _output = intermediate_steps[-1]
+        input = None
+        output = ""
+        tool_progress_messages: list[AssistantMessage] = []
+
+        try:
+            input = FilterOptionsTool.model_validate({"name": action.tool, "arguments": action.tool_input})
+        except ValidationError as e:
+            output = str(
+                ChatPromptTemplate.from_template(REACT_PYDANTIC_VALIDATION_EXCEPTION_PROMPT, template_format="mustache")
+                .format_messages(exception=e.errors(include_url=False))[0]
+                .content
+            )
+        else:
+            # First check if we've reached the terminal stage and return the filter options
+            if input.name == "final_answer":
+                try:
+                    # Extract the full response structure
+                    full_response = {
+                        "result": input.arguments.result,  # type: ignore
+                        "data": input.arguments.data,  # type: ignore
+                    }
+
+                    return PartialAssistantState(
+                        generated_filter_options=full_response,
+                        change=state.change,
+                        intermediate_steps=[],
+                        messages=[
+                            AssistantToolCallMessage(
+                                tool_call_id=state.root_tool_call_id or "",  # Default to empty string if None
+                                content=json.dumps(full_response),
+                            )
+                        ],
+                    )
+                except Exception as e:
+                    # Fall through to handle as regular tool call
+                    output = f"Error processing final answer: {e}"
+
+            # The agent has requested help, so we return a message to the root node
+            if input.name == "ask_user_for_help":
+                help_message = input.arguments.request  # type: ignore
+                return self._get_reset_state(str(help_message), input.name, state)
+
+        # If we're still here, check if we've hit the iteration limit within this cycle
+        if len(intermediate_steps) >= self.MAX_ITERATIONS:
+            return self._get_reset_state(FILTER_OPTIONS_ITERATION_LIMIT_PROMPT, "max_iterations", state)
+
+        if input and not output:
+            # Generate progress message before executing tool
+
+            if input.name == "retrieve_entity_property_values":
+                entity = getattr(input.arguments, "entity", "unknown")
+                property_name = getattr(input.arguments, "property_name", "unknown")
+                tool_progress_messages.append(
+                    AssistantMessage(
+                        content=f"🔍 Fetching values for {entity} property '{property_name}'...", id=str(uuid4())
+                    )
+                )
```

Comment:
This way of exposing data exploration progress is much different from what we do in insights:

<img width="356" height="194" alt="Screenshot 2025-07-12 at 00 45 05" src="https://github.com/user-attachments/assets/defbdc02-c33b-459a-a435-5d13f21e25e6" />

vs.

<img width="322" height="87" alt="Screenshot 2025-07-12 at 00 45 36" src="https://github.com/user-attachments/assets/6f636721-e947-46f3-b64e-1768e7bcbb64" />

Can we use the insights-style `ReasoningMessage.substeps` here?